Thermostat remove deprecated config

MrMDavidson opened the pull request:
# What does this implement/fix?

Following on from #3298 this PR moves any deprecated configuration from raise an error with descriptive messaging on how to convert between the old configuration and the new configuration. The intent is to have this merged, released, and then the following release remove all the deprecated configuration handling code.

Secondly, as part of this, it introduces `default_preset` as the successor for `default_mode`. This will change the preset on start up. This effectively allows the control of not just default modes, but default temperatures, swing modes, and fan modes on the thermostat.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2292

## Breaking change notes

`default_mode` has been replaced with `default_preset`. See [the docs](https://next.esphome.io/components/climate/thermostat.html#default-preset) for more detail.

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example legacy yaml
climate:
  - platform: thermostat
    name: My Thermostat
    default_mode: Off
    away_config:
        default_target_temperature_low: 16
        default_target_temperature_high: 30
    default_target_temperature_low: 20
    default_target_temperature_high: 24
```

```yaml
# Example new yaml
climate:
  - platform: thermostat
    name: My Thermostat
    default_preset: Startup Preset
    startup_behavior: APPLY_DEFAULT_PRESET
    preset:
      - name: Away
        default_target_temperature_low: 16
        default_target_temperature_high: 30
      - name: Home
        default_target_temperature_low: 20
        default_target_temperature_high: 24
      - name: Startup Preset
        mode: off
        default_target_temperature_low: 20
        default_target_temperature_high: 24
    
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

TBD

MrMDavidson committed:
Raise errors for all the now deprecated options

MrMDavidson committed:
Fix CONF_DEFAULT_PRESET detection

MrMDavidson committed:
Stop attempting to set the non-existent normal_config

MrMDavidson committed:
Add support for default presets

MrMDavidson commented:
Opening this now for feedback. The two questions are;
1. Do we want to do all the configuration erroring / directing to the new configs? (It's already done, so may as well leave it)
2. With the start up mode.... in doing this I'm not convinced save/restore works correctly for the Thermostat. I've always had issues with it occasionally putting my thermostat into invalid modes. And looking at it the base Climate doesn't really allow extension. Is this something we should extend?
  a. Following on from this; currently because of the above the default preset is applied every boot even if the state was attempted to restore. Thoughts?

kbx81 commented:
I use this exclusively with ESP32s on my thermostat boards and I've never once had an (unexpected) issue with it restoring the mode and set points to exactly how they were prior to a reset or power loss. I hate to throw out a "it works on my machine" but it does (and quite nicely I might add). I'd much prefer it if we didn't break this if at all possible -- perhaps (for now) the right approach is to just ignore (re)setting the preset if other data (mode, set points) is/can be restored and, if this cannot be restored, _then_ apply the default preset (and all related settings, of course).

Of course the longer-winded approach would be to update `climate` to do what's required for fully, proper support of the preset situation but this may be a bit more involved since it has the potential to break other things that rely on it.

Sooo that all said, I (once again) don't have a strong preference because I just don't really use the preset functionality. I set it up, it's neat, I like it, but I just don't use it. 😄 With respect to restoring settings, I just don't want to break something that isn't already broken (at least, not broken on ESP32, haha). 😇

MrMDavidson committed:
Fix correct detection of Two Point temperature mode

MrMDavidson committed:
Fix lint issues

MrMDavidson committed:
Fix tests

MrMDavidson committed:
Generate correct yaml for equivalent configurations

MrMDavidson committed:
Remove debug code

MrMDavidson committed:
Only set default preset if the thermostat does not have state to restore

MrMDavidson commented:
> I use this exclusively with ESP32s on my thermostat boards and I've never once had an (unexpected) issue with it restoring the mode and set points to exactly how they were prior to a reset or power loss. I hate to throw out a "it works on my machine" but it does (and quite nicely I might add).

Actually, now I think about it, the issue I may have been facing was that for new devices the fan modes, etc, may not get sent to sensible defaults... the thermostat would then reapply the config `min_fan_mode_switching_time`. So it's possible that the issue I was thinking of is actually (hopefully) solved by the introduction of a default preset where the user can set fan modes, etc, as a result.

I've made a change in line with your suggestion of only setting the default if a restore state doesn't exist.

> Of course the longer-winded approach would be to update climate to do what's required for full, proper support of the preset situation but this may be a bit more involved since it has the potential to break other things that rely on it.

I haven't looked closely... but it may be possible this should work. I'm just not 100% sure at what point state restore is meant to apply? If I do an OTA update, is the state persisted? Or is it only for power outages, etc?

kbx81 commented:
> I haven't looked closely... but it may be possible this should work. I'm just not 100% sure at what point state restore is meant to apply? If I do an OTA update, is the state persisted? Or is it only for power outages, etc?


I'm not sure I know the specific answer to your question, but I guess my expectation would be that, anytime the ESP resets/reboots, the state is restored. I guess I don't see an alternative behavior that makes sense. For example, you asked about OTA...would you not restore the state because of an OTA update and just assume garbage values? That doesn't seem like it would make sense...but maybe I'm misunderstanding what you were trying to convey. 😇

kbx81 commented:
@MrMDavidson where did you go? I miss you 😄

I had another thought on this -- maybe we should add a flag that allows the user to define whether to "force preset values at start-up" or "restore from RTC memory at start-up". I think the default (for compatibility) would be to use the restore data and only force the set points as defined by the default preset if the user enables the flag. 😄 Any thoughts?

MrMDavidson commented:
Whoops! The notification for this was put in my "Deal with this when I get home" pile and then life happened... twice... 😂

> I'm not sure I know the specific answer to your question, but I guess my expectation would be that, anytime the ESP resets/reboots, the state is restored. I guess I don't see an alternative behavior that makes sense. For example, you asked about OTA...would you not restore the state because of an OTA update and just assume garbage values? That doesn't seem like it would make sense...but maybe I'm misunderstanding what you were trying to convey.

Given what you've said about restoring on shutdown, which appears to be the current behaviour, it makes me question what the purpose of the `default_mode` option is even for. It'd only get used on _first_ boot with a new "version" of the config (defined by when the underlying `Climate` component changed its salt value). In which case are we better off just nuking this option instead of porting it to presets?

> I had another thought on this -- maybe we should add a flag that allows the user to define whether to "force preset values at start-up" or "restore from RTC memory at start-up". I think the default (for compatibility) would be to use the restore data and only force the set points as defined by the default preset if the user enables the flag. 😄 Any thoughts?

That'd be the other option. I'd be curious as to if users would ever want that behaviour though. As an embedded device when something goes "wrong" (eg. Reboot) I think the vast majority of users would want it to restore to its original state. The `Thermostat` component already has the `startup_delay` type options for protection of repeated cycling.

From my perspective we have a few options;
1. Remove `default_mode`. Pros: Removes limited use case option, Cons: Some loss of corner case functionality
2. Port `default_mode` to `default_preset`. Pros: Keep corner case but with expanded power, Cons: Limited use case
3. Add On-start-up-do toggle between memory/configured preset. Pros: Most flexible, Cons: More code, limited use case on a limited use case.

My preference is 1 or 2. Thoughts?

kbx81 commented:
Yeah, these are all pretty edge-case-y, but I've learned from past conversations that there are a few folks who prefer that it sets/reverts to default values after a reboot (as opposed to just restoring them...which is (clearly) what I--and apparently most other folks--expect of it. haha).

I had originally added the existing `default_mode` option in a (quick, hasty) effort to address some confusion that surfaced about how it would behave when a new/modified config was installed (something that would otherwise prevent it from restoring from memory). That said, I don't have strong feelings about it because it is not a situation that's ever impacted me, but it's still a thing that has come up at least a couple times that I can recall. I also am kind of under the impression that restoring works better/more reliably with ESP32s as opposed to '8266s...but I don't have as much experience with '8266s because I avoid them as much as possible in favor of the ESP32. haha

I do tend to agree with what you said above -- I think your option 2 makes the most sense given the general direction. That said, I do still think it would be a worthwhile endeavor to add a flag to disable restoring from memory (maybe something like `always_restore_default_preset`) -- I'm quite sure there exists (at very least) a small number of users who would like to have this option available. 😄

MrMDavidson committed:
Add restore_default_preset_on_boot option
If set to True then the default_preset will be applied on every boot. If False (Default) state will be restored from memory as per prior versions

MrMDavidson commented:
> Yeah, these are all pretty edge-case-y, but I've learned from past conversations that there are a few folks who prefer that it sets/reverts to default values after a reboot (as opposed to just restoring them...which is (clearly) what I--and apparently most other folks--expect of it. haha).
> 
> I had originally added the existing `default_mode` option in a (quick, hasty) effort to address some confusion that surfaced about how it would behave when a new/modified config was installed (something that would otherwise prevent it from restoring from memory). That said, I don't have strong feelings about it because it is not a situation that's ever impacted me, but it's still a thing that has come up at least a couple times that I can recall. I also am kind of under the impression that restoring works better/more reliably with ESP32s as opposed to '8266s...but I don't have as much experience with '8266s because I avoid them as much as possible in favor of the ESP32. haha
> 
> I do tend to agree with what you said above -- I think your option 2 makes the most sense given the general direction. That said, I do still think it would be a worthwhile endeavor to add a flag to disable restoring from memory (maybe something like `always_restore_default_preset`) -- I'm quite sure there exists (at very least) a small number of users who would like to have this option available. 😄

Okay, I've introduced a new `restore_default_preset_on_boot` option. I'm not convinced with the name/behaviour. Perhaps `apply_default_preset_on_boot` would be clearer, rather than `restore_` ? The other option might be `default_preset_behaviour` as an enum of `restore_memory` and `from_config` (or something) which allows for further extension?

MrMDavidson committed:
Apply lint suggestions

kbx81 commented:
> Perhaps `apply_default_preset_on_boot` would be clearer, rather than `restore_`

I don't have a strong opinion here -- I think as long as it's explained clearly in the docs it should be fine. I'm not sure the complexity of an enum is really necessary, although it's not a bad idea. I just can't think of how/why we would need to extend it beyond either of the two options.

MrMDavidson commented:
> > Perhaps `apply_default_preset_on_boot` would be clearer, rather than `restore_`
> 
> I don't have a strong opinion here -- I think as long as it's explained clearly in the docs it should be fine. I'm not sure the complexity of an enum is really necessary, although it's not a bad idea. I just can't think of how/why we would need to extend it beyond either of the two options.

Yeah, I also can't really think of any other potential options. That said the advantage of the enum approach is it's a bit more self documenting. `restore_preset_on_boot: false` isn't immediately obvious as to what it _would_ do just that it wouldn't restore it. Same as `apply_default_preset_on_boot: false`. Compared with `default_preset_behaviour: restore_from_memory` and `default_preset_behaviour: apply_from_config` (or whatever) is more self evident as to what they're doing in both scenarios. 🤔

kbx81 commented:
Yeah, I see what you're saying. Perhaps we call the parameter `startup_behavior` and the values are `restore_from_memory` (default) and `apply_default_preset`. Does that seem like it would be clear enough?

MrMDavidson committed:
Switch from restore_default_preset_on_boot to an enum for startup_behavior
This gives better self-documentation as well as the option for extending to other options down the track

MrMDavidson commented:
> Yeah, I see what you're saying. Perhaps we call the parameter `startup_behavior` and the values are `restore_from_memory` (default) and `apply_default_preset`. Does that seem like it would be clear enough?

I like it - done. Any suggestions for an appropriate synonym for behaviour/behavior that doesn't suffer from spelling differences across English locales? `startup_mode` ? `startup_action` ? 🤔

MrMDavidson commented:
> With the start up mode.... in doing this I'm not convinced save/restore works correctly for the Thermostat. I've always had issues with it occasionally putting my thermostat into invalid modes. And looking at it the base Climate doesn't really allow extension. Is this something we should extend?

Going back to this; I think I've found the source of this. It seems that a custom preset can't be restored. If you use one of the built in ones (Tested with Home) it works fine. I think this might have to be a limitation of the component until we move to having derived implementations of `Climate` allow their own state to be restored/saved?

MrMDavidson committed:
Lint fixes

kbx81 commented:
> I like it - done. Any suggestions for an appropriate synonym for behaviour/behavior that doesn't suffer from spelling differences across English locales? startup_mode ? startup_action ? 🤔

One Idea I had: `on_boot_restore_from:` -- `memory` (default) / `default_preset`

MrMDavidson committed:
Rename startup_behavior to on_boot_restore_from
This removes any issues with different English locales

MrMDavidson commented:
> > I like it - done. Any suggestions for an appropriate synonym for behaviour/behavior that doesn't suffer from spelling differences across English locales? startup_mode ? startup_action ? 🤔
> 
> One Idea I had: `on_boot_restore_from:` -- `memory` (default) / `default_preset`

That works for me. Done!

If it's all good I'll update the docs associated with this change

kbx81 commented:
Should be fine -- document away! 😄

MrMDavidson commented:
Documentation added! https://github.com/esphome/esphome-docs/pull/2292

kbx81 commented:
Giving https://github.com/esphome/issues/issues/3525 a mention here...probably falls into the "won't fix" category but I thought I'd link it nonetheless. 😄

kbx81 commented:
@MrMDavidson please have a look at [the issue](https://github.com/esphome/issues/issues/3525) when you have a moment -- maybe we could squeeze a small PR in to fix this prior to this being released (I don't think this specific PR will make it into this release...timing...but next month I'd like to get it in for sure!)?

kbx81 commented:
Are you good with this @MrMDavidson? I'll try to do a more thorough review over the next couple of days so we can get it into the October release. 🙂

MrMDavidson commented:
> Are you good with this @MrMDavidson? I'll try to do a more thorough review over the next couple of days so we can get it into the October release. 🙂

Yup, from my perspective it looks good. Would be keen to have some other eyes/users on it though to see how it plays out for them.

kbx81 committed:
Fix comparable_preset yaml output alignment

kbx81 committed:
Add dump of on_boot_restore_from setting
